### ISSUE TITLE
Persistence in Browser / WASM

### ISSUE THREAD

**Issue by slater1** [praeclarum/sqlite-net#1173]
I'm trying to use sqlite-net in an Avalonia browser app. I've wired up sqlite-net and it works on Desktop, iOS, and Android, but in the browser, nothing seems to be persisted to `localStorage` or `sessionStorage`. All saved data is lost on page refresh, as if no sqlite db was saved.

sqlite documents some persistence options [here](https://sqlite.org/wasm/doc/tip/persistence.md#jsstoragedb). How can I access these from sqlite-net?

Thanks in advance.

**Comment by slater1** [comment]
Adding a little more detail. I've tried both of the following connection strings

`file:local?vfs=kvvfs`
`file:local?vfs=opfs`

per the docs linked in the first post. Although e.g. creation of the tables and inserts appear to succeed (i.e. insert a row returns `1`), nothing shows in the developer console under Storage => Local Storage, and everything is forgotten after a page refresh.

Do I need to use sqlite3.js and sqlite3.wasm manually i.e. with .NET 7 JS interop?

**Comment by ericsink** [comment]
Just a quick reply to explain why it might be tricky for anyone to respond with help for you.

I suspect you're using the version of sqlite-net which depends on SQLitePCLRaw, which I maintain.

SQLitePCLRaw has bundle packages which contain builds of sqlite for various platforms.

One of those builds is Wasm, but that build was provided by a contributor pull request, and I don't know much about it.

**Comment by slater1** [comment]
Hey Eric, and thanks for replying. Indeed I'm using your library (thank you for it). The specific references I'm using are

```xml
    <PackageReference Include="sqlite-net-pcl" Version="1.8.116" />
    <PackageReference Include="SQLitePCLRaw.bundle_green" Version="2.1.5" />
```

Which works on all platforms, except on Browser where DB operations persist until a page refresh.

What I was hoping to find was something like

```xml
    <PackageReference Include="SQLitePCLRaw.provider.wasm" Version="2.1.5" />-->
```

Alas, it seems my options are either use SQLitePCLRaw.provider.sqlite3" and then somehow get a WASM-compatible non-amnesiac sqlite3.dll into my app's address space (as it seems [Uno has done](https://github.com/unoplatform/uno.sqlite-wasm)), which at the moment is over my head, or to implement my own `ISQLite3Provider` which uses .NET7 interop and sql.js as the backend, which seems circuitous but achievable. Do you see other options?

There's a lot of examples floating around for Blazor, but I'm using Avalonia.

Thanks again

**Comment by slater1** [comment]
I figured it out. Of the seemingly myriad options, [this post ](https://www.thinktecture.com/blazor/ef-core-and-sqlite-in-browser/) is what I went with. 

I neither wrote a custom ISQLite3Provider nor found a different sqlite3 wasm .dll/.so. The wasm binary provided by SQLitePCLRaw was working, and by default it writes to the Emscripten `MEMFS` filesystem, which is ephemeral and lost on a page refresh.

After following the blog post and a bit of config tweaking for .NET 7 , my Avalonia WASM app is now writing to `IDBFS` (IndexedDB File System), i.e. emulating a filesystem using IndexedDB, and DB changes persist after a refresh.

The tweak for .NET 7 was, compared to .NET 6 (for which the linked blog post was written), .NET 7 [exports ES6 modules differently](https://github.com/dotnet/runtime/pull/66868/files), making it necessary to add the emcc (Emscripten compiler frontend) flag `EXPORTED_RUNTIME_METHODS`.

```xml
    <!-- Enable Emscripten support for IDBFS. We only care about IDBFS but include the others because we can't append to EXPORTED_RUNTIME_METHODS -->
    <EmccExtraLDFlags>-lidbfs.js -s 'EXPORTED_RUNTIME_METHODS=["IDBFS", "FS", "setValue", "UTF8ToString", "UTF8ArrayToString", "ccall", "cwrap"]'</EmccExtraLDFlags>
```

There's a slightly shorter way to do it but the emcc flag is deprecated since 2021
```xml
    <!-- EXTRA_EXPORTED_RUNTIME_METHODS generates a deprecated warning-->
    <!-- See https://github.com/emscripten-core/emscripten/pull/13895/files-->
    <EmccExtraLDFlags>-lidbfs.js -s 'EXTRA_EXPORTED_RUNTIME_METHODS=["IDBFS"]'</EmccExtraLDFlags>

```

Performance: Each sync writes or reads the whole DB to/from memory, which isn't going to scale. There's already [been a discussion](https://github.com/ericsink/SQLitePCL.raw/discussions/440) about that.
